### PR TITLE
[build] Warning Suppression PR #1: Ignored selected compilation warnings & Fixed clang-specific warnings

### PR DESCRIPTION
**Pull request by jim19930609**
Related issue = #3335, #2927

Changes in compiler options:
1. -Wno-ignored-attributes
[Global] By default, CXX compiler will throw a warning if it decides to ignore an attribute, for example "[[ maybe unused ]]".
However, this behaviour diverges across different compilers (GCC/CLANG), as well as different compiler versions.
Therefore we disable such warnings for now.

2. -Wno-nullability-completeness
[Global] Clang warns if a C++ pointer's nullability wasn't marked explicitly (__nonnull, nullable, ...).
Nullability seems to be a clang-specific feature, thus we disable this warning.

3. -Wno-unused-private-field
[Global] Disable warning for unused-private-field for convenience in development.

4. -Wno-unneeded-internal-declaration
[Global] By evaluating "constexpr", compiler throws a warning for functions known to be dead at compile time.
However, some of these "constexpr" are debug flags and will be manually enabled upon debuging.

### CODE REVIEW

File: taichi/backends/vulkan/vulkan_utils.cpp [hunk 1]
```diff
@@ -5,31 +5,7 @@
...
-  vkEnumerateDeviceExtensionProperties(physicalDevice, kNoLayerName, &count,
-                                       nullptr);
-  std::vector<VkExtensionProperties> extensions(count);
-  vkEnumerateDeviceExtensionProperties(physicalDevice, kNoLayerName, &count,
-                                       extensions.data());
-  return extensions;
-}
-
-}  // namespace
+namespace {}  // namespace
```
Comment (k-ye) [inline taichi/backends/vulkan/vulkan_utils.cpp:8]: Feel free to remove this if it's no longer in use!

Comment (wyushun) [inline taichi/backends/vulkan/vulkan_utils.cpp:8]: it's an empty file?

Comment (jim19930609) [reply taichi/backends/vulkan/vulkan_utils.cpp:8]: yup, let me remove that

**Review by ailzhang — APPROVED** [review]
LGTM!

**Pull request merged**